src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java

```diff
@@ -116,11 +116,18 @@ public JavaAstVisitor(CommonTokenStream tokenStream) {
 
     @Override
     public DetailAstImpl visitCompilationUnit(JavaLanguageParser.CompilationUnitContext ctx) {
-        final DetailAstImpl root = visit(ctx.children.get(0));
-        for (int i = 1; i < ctx.children.size(); i++) {
-            addLastSibling(root, visit(ctx.children.get(i)));
+        final DetailAstImpl compilationUnit;
+        // 'EOF' token is always present; therefore if we only have one child, we have an empty file
+        final boolean isEmptyFile = ctx.children.size() == 1;
+        if (isEmptyFile) {
+            compilationUnit = null;
         }
-        return root;
+        else {
+            compilationUnit = createImaginary(TokenTypes.COMPILATION_UNIT);
+            // last child is 'EOF', we do not include this token in AST
```

Comment:
Why do we even have EOF in the tree if its not displayed/used? Can we remove it in another issue?

Reply:
`EOF` is part of every input, and it is standard practice to include this in any ANTLR4 grammar. This signals to the parser that it has completed it's task.

Reply:
Oh, so it is something ANTLR adds, not us?

Reply:
It is possible to remove it, but generally not a good idea. I am not sure how ANTLR determines we are at EOF (the actual `EOF` character as a file terminator has not been used for some time now), perhaps it counts bytes, but I have seen several reports of undefined behavior without using the `EOF` token at the end of the start rule.

Reply:
Ok, then I will resolve this as well and we will not touch it.